utils/syncstream/syncstream.go

```diff
@@ -0,0 +1,132 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+// Copyright (C) 2022 EPAM Systems, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package syncstream
+
+import (
+	"context"
+	"errors"
+	"reflect"
+	"sync"
+)
+
+/***********************************************************************************************************************
+ * Types
+ **********************************************************************************************************************/
+
+type SyncStream struct {
+	sync.Mutex
+	requestID int
+	requests  []request
+}
+
+type request struct {
+	id           int
+	responseType reflect.Type
+	channel      chan<- interface{}
+}
+
+/***********************************************************************************************************************
+ * Vars
+ **********************************************************************************************************************/
+
+var ErrTimeout = errors.New("wait for response timeout")
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+func New() *SyncStream {
+	return &SyncStream{}
+}
+
+func (stream *SyncStream) Send(ctx context.Context, sendFunc func() error,
+	responseType reflect.Type,
+) (response interface{}, err error) {
+	stream.Lock()
+
+	channel := make(chan interface{}, 1)
+
+	request := request{
+		id:           stream.requestID,
+		responseType: responseType,
+		channel:      channel,
+	}
+
+	stream.requests = append(stream.requests, request)
+
+	stream.requestID++
```

Comment:
what if we generate a random value instead of an increment approach?

I mean it's a rare case, but we can overflow the value, cause we increment but not decrement, and theoretically if the application will be working for a long time it's possible to get an overflow 

but up to you